File: src-tauri/experts/skills/iyw-image-workflows/scripts/iyw_knowledge.py
#!/usr/bin/env python3
"""Standalone CLI for the captured IYW knowledge search API."""

from __future__ import annotations

import argparse
import asyncio
import json
import sys
from typing import Any

from iyw_image import IywClient, IywError, _add_connection_args, _resolve_token

KNOWLEDGE_PREFIX = "/ai-agent-new/api/knowledge"
DEFAULT_LIMIT = 10
DEFAULT_DENSE_WEIGHT = 0.5


def build_payload(args: argparse.Namespace) -> dict[str, Any]:
    query = str(args.query or "").strip()
    if not query:
        raise IywError("knowledge query is required", "invalid_input")
    if args.limit < 1:
        raise IywError("--limit must be greater than zero", "invalid_input")
    if not 0 <= args.dense_weight <= 1:
        raise IywError("--dense-weight must be between 0 and 1", "invalid_input")
    return {
        "category": args.category,
        "query": query,
        "folderId": args.folder_id,
        "fileId": args.file_id,
        "limit": args.limit,
        "denseWeight": args.dense_weight,
    }


def _normalized_chunk(value: Any) -> dict[str, Any]:
    if not isinstance(value, dict):
        raise IywError(
            "knowledge search returned an invalid result item", "invalid_response"
        )
    document = value.get("doc_info")
    if document is not None and not isinstance(document, dict):
        raise IywError(
            "knowledge search returned invalid document metadata", "invalid_response"
        )
    document = document or {}
    return {
        "id": value.get("id"),
        "score": value.get("score"),
        "content": value.get("content"),
        "md_content": value.get("md_content"),
        "chunk_type": value.get("chunk_type"),
        "doc_id": document.get("doc_id"),
        "doc_name": document.get("doc_name"),
        "doc_type": document.get("doc_type"),
    }


def normalize_search_response(data: dict[str, Any]) -> dict[str, Any]:
    result = data.get("result") if isinstance(data, dict) else None
    if not isinstance(result, dict):
        raise IywError("knowledge search omitted result", "invalid_response")
    if result.get("code") != 0:
        raise IywError(
            str(result.get("message") or "knowledge search failed"),
            "knowledge_search_failed",
        )
    result_data = result.get("data")
    if not isinstance(result_data, dict):
        raise IywError("knowledge search omitted result data", "invalid_response")
    items = result_data.get("result_list")
    count = result_data.get("count")
    if not isinstance(items, list) or not isinstance(count, int) or count < 0:
        raise IywError(
            "knowledge search returned an unexpected shape", "invalid_response"
        )
    return {
        "count": count,
        "results": [_normalized_chunk(item) for item in items],
    }


def _client(args: argparse.Namespace) -> IywClient:
    if args.dry_run:
        return IywClient(
            args.base_url,
            "",
            KNOWLEDGE_PREFIX,
            args.timeout,
            allow_missing_token=True,
        )
    return IywClient(
        args.base_url,
        _resolve_token(args.token),
        KNOWLEDGE_PREFIX,
        args.timeout,
    )


async def run_search(args: argparse.Namespace) -> dict[str, Any]:
    payload = build_payload(args)
    response = await _client(args).request(
        "search", payload, dry_run=args.dry_run
    )
    return response if args.dry_run else normalize_search_response(response)


def build_parser() -> argparse.ArgumentParser:
    parser = argparse.ArgumentParser(description="Standalone IYW knowledge search CLI")
    subparsers = parser.add_subparsers(dest="command", required=True)
    search = subparsers.add_parser("search", help="search the IYW knowledge base")
    _add_connection_args(search)
    search.add_argument("--query", required=True)
    search.add_argument("--category", type=int, default=0)
    search.add_argument("--folder-id", type=int)
    search.add_argument("--file-id")
    search.add_argument("--limit", type=int, default=DEFAULT_LIMIT)
    search.add_argument(
        "--dense-weight", type=float, default=DEFAULT_DENSE_WEIGHT
    )
    return parser


def _failure(exc: Exception) -> int:
    code = exc.code if isinstance(exc, IywError) else "invalid_input"
    retryable = exc.retryable if isinstance(exc, IywError) else False
    print(
        json.dumps(
            {
                "ok": False,
                "error": {
                    "code": code,
                    "message": str(exc),
                    "retryable": retryable,
                },
            },
            ensure_ascii=False,
            indent=2,
        )
    )
    return 1


def _configure_stdout() -> None:
    reconfigure = getattr(sys.stdout, "reconfigure", None)
    if callable(reconfigure):
        reconfigure(encoding="utf-8")


def main(argv: list[str] | None = None) -> int:
    _configure_stdout()
    args = build_parser().parse_args(argv)
    try:
        result = asyncio.run(run_search(args))
    except (IywError, OSError, TypeError, ValueError) as exc:
        return _failure(exc)
    print(json.dumps({"ok": True, "data": result}, ensure_ascii=False, indent=2))
    return 0


if __name__ == "__main__":
    raise SystemExit(main())

File: src-tauri/src/acp/builtin_mcp/iyw_knowledge.rs
use serde::Deserialize;
use serde_json::{json, Map, Value};

use super::iyw_service::IywGatewayService;

#[derive(Debug, Deserialize)]
#[serde(rename_all = "camelCase", deny_unknown_fields)]
struct KnowledgeRequest {
    query: String,
    #[serde(default)]
    category: i64,
    folder_id: Option<i64>,
    file_id: Option<String>,
    #[serde(default = "default_limit")]
    limit: i64,
    #[serde(default = "default_dense_weight")]
    dense_weight: f64,
}

const fn default_limit() -> i64 {
    10
}

const fn default_dense_weight() -> f64 {
    0.5
}

pub(super) async fn search(
    service: &IywGatewayService,
    arguments: Value,
) -> Result<rmcp::model::CallToolResult, rmcp::ErrorData> {
    let request: KnowledgeRequest = serde_json::from_value(arguments)
        .map_err(|error| rmcp::ErrorData::invalid_params(error.to_string(), None))?;
    validate(&request)?;
    let body = json!({
        "category": request.category,
        "query": request.query.trim(),
        "folderId": request.folder_id,
        "fileId": request.file_id,
        "limit": request.limit,
        "denseWeight": request.dense_weight,
    });
    let value = service
        .post_gateway("/ai-agent-new/api/knowledge", "search", body)
        .await?;
    let result = normalize(value)?;
    Ok(rmcp::model::CallToolResult::structured(result))
}

fn validate(request: &KnowledgeRequest) -> Result<(), rmcp::ErrorData> {
    if request.query.trim().is_empty() || request.query.chars().count() > 4096 {
        return Err(rmcp::ErrorData::invalid_params(
            "query must contain between 1 and 4096 characters",
            None,
        ));
    }
    if !(1..=100).contains(&request.limit) {
        return Err(rmcp::ErrorData::invalid_params(
            "limit must be between 1 and 100",
            None,
        ));
    }
    if !(0.0..=1.0).contains(&request.dense_weight) {
        return Err(rmcp::ErrorData::invalid_params(
            "denseWeight must be between 0 and 1",
            None,
        ));
    }
    Ok(())
}

fn normalize(value: Value) -> Result<Value, rmcp::ErrorData> {
    let result = value
        .get("result")
        .and_then(Value::as_object)
        .ok_or_else(|| invalid_response("knowledge search omitted result"))?;
    if result.get("code").and_then(Value::as_i64) != Some(0) {
        return Err(invalid_response("knowledge search failed"));
    }
    let data = result
        .get("data")
        .and_then(Value::as_object)
        .ok_or_else(|| invalid_response("knowledge search omitted result data"))?;
    let count = data
        .get("count")
        .and_then(Value::as_i64)
        .filter(|count| *count >= 0)
        .ok_or_else(|| invalid_response("knowledge search returned invalid count"))?;
    let items = data
        .get("result_list")
        .and_then(Value::as_array)
        .ok_or_else(|| invalid_response("knowledge search returned invalid results"))?;
    let results = items
        .iter()
        .map(normalize_item)
        .collect::<Result<Vec<_>, _>>()?;
    Ok(json!({"count": count, "results": results}))
}

fn normalize_item(value: &Value) -> Result<Value, rmcp::ErrorData> {
    let object = value
        .as_object()
        .ok_or_else(|| invalid_response("knowledge result item is invalid"))?;
    let document = object
        .get("doc_info")
        .and_then(Value::as_object)
        .cloned()
        .unwrap_or_else(Map::new);
    Ok(json!({
        "id": object.get("id"),
        "score": object.get("score"),
        "content": object.get("content"),
        "md_content": object.get("md_content"),
        "chunk_type": object.get("chunk_type"),
        "doc_id": document.get("doc_id"),
        "doc_name": document.get("doc_name"),
        "doc_type": document.get("doc_type"),
    }))
}

fn invalid_response(message: &'static str) -> rmcp::ErrorData {
    rmcp::ErrorData::internal_error(message, None)
}
